File: elixir/raspi_tank/lib/tcp_tank.ex
defmodule TcpTank do
  alias RaspiTank.Server

  def start(port, left_pin, right_pin) do
    {:ok, lsock} = :gen_tcp.listen(port, [:binary, {:packet, 0}, 
                                         {:active, false}])
    do_accept(lsock, left_pin, right_pin)
  end

  def do_accept(lsock, left_pin, right_pin) do
    {:ok, sock} = :gen_tcp.accept(lsock)
    {:ok, pid} = Server.start([left_pin, right_pin])
    do_recv(sock, pid)
    do_accept(lsock, left_pin, right_pin)
  end

  def do_recv(sock, pid) do
    case :gen_tcp.recv(sock, 0) do
      {:ok, data} ->
        handle_data(data, pid)
        do_recv(sock, pid)
      {:error, :closed} ->
        :ok
    end
  end

  def handle_data(data, pid) do
    {left, right} = extract_values(data)
    pid |> Server.update(left, -1*right)
    pid |> Server.blast
  end

  def extract_values(data) do
    {:match, [left, right]} = :re.run(data, value_extractor(), [{:capture, ['LEFT', 'RIGHT'], :list}])
    {input_to_num(left), input_to_num(right)}
  end

  def value_extractor() do
    "L(?<LEFT>[0-9.\-]+)R(?<RIGHT>[0-9.\-]+).*"
  end

  def input_to_num(n) do
    case :string.to_float(n) do
      {:error,:no_float} -> list_to_integer(n)
      {f,_rest} -> f
    end
  end
end

File: erlang/raspi_tank/examples/tcp_tank.erl
-module(tcp_tank).
-export([start/3]).

start(Port, LeftPin, RightPin) ->
  {ok, LSock} = gen_tcp:listen(Port, [binary, {packet, 0},
                                      {active, false}]),
  do_accept(LSock, LeftPin, RightPin).

do_accept(LSock, LeftPin, RightPin) ->
  {ok, Sock} = gen_tcp:accept(LSock),
  {ok, Pid} = tank_server:start(LeftPin, RightPin),
  do_recv(Sock, Pid),
  do_accept(LSock, LeftPin, RightPin).

do_recv(Sock, Pid) ->
  case gen_tcp:recv(Sock, 0) of
    {ok, Data} ->
      handle_data(Data, Pid),
      do_recv(Sock, Pid);
    {error, closed} ->
      ok
  end.

handle_data(Data, Pid) ->
  {Left, Right} = extract_values(Data),
  Pid ! {update, Left, -1*Right},
  Pid ! blast.

extract_values(Data) ->
  {match, [Left, Right]} = re:run(Data, value_extractor(), [{capture, ['LEFT', 'RIGHT'], list}]),
  io:format("left: ~p, right: ~p\n", [Left, Right]),
  {input_to_num(Left), input_to_num(Right)}.

value_extractor() ->
  "L(?<LEFT>[0-9.\-]+)R(?<RIGHT>[0-9.\-]+).*".

input_to_num(N) ->
  case string:to_float(N) of
    {error,no_float} -> list_to_integer(N);
    {F,_Rest} -> F
  end.
